### ISSUE TITLE
Outbound SOAP over HTTPS broken in 2.0.4

### ISSUE THREAD

**Issue by candlerb** [zatosource/zato#454]
SOAP HTTPS connections are broken in 2.0.4, I am guessing that the fix for #446 has broken it.

~~~
  File ...
    resp = zato.outgoing.soap.get(chan).conn.send(zato.cid, body)
  File "/opt/zato/2.0.4/zato-server/src/zato/server/connection/http_soap/outgoing.py", line 319, in post
    return self.http_request('POST', cid, data, params, *args, **kwargs)
  File "/opt/zato/2.0.4/zato-server/src/zato/server/connection/http_soap/outgoing.py", line 295, in http_request
    response = self.invoke_http(cid, method, address, data, headers, {}, params=qs_params, *args, **kwargs)
  File "/opt/zato/2.0.4/zato-server/src/zato/server/connection/http_soap/outgoing.py", line 79, in invoke_http
    cert=cert, verify=verify, timeout=self.config['timeout'], *args, **kwargs)
  File "/opt/zato/2.0.4/eggs/requests-2.6.0-py2.7.egg/requests/sessions.py", line 464, in request
    resp = self.send(prep, **send_kwargs)
  File "/opt/zato/2.0.4/eggs/requests-2.6.0-py2.7.egg/requests/sessions.py", line 579, in send
    r = adapter.send(request, **kwargs)
  File "/opt/zato/2.0.4/eggs/requests-2.6.0-py2.7.egg/requests/adapters.py", line 370, in send
    timeout=timeout
  File "/opt/zato/2.0.4/eggs/requests-2.6.0-py2.7.egg/requests/packages/urllib3/connectionpool.py", line 538, in urlopen
    body=body, headers=headers)
  File "/opt/zato/2.0.4/eggs/requests-2.6.0-py2.7.egg/requests/packages/urllib3/connectionpool.py", line 343, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/lib/python2.7/httplib.py", line 973, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib/python2.7/httplib.py", line 1007, in _send_request
    self.endheaders(body)
  File "/usr/lib/python2.7/httplib.py", line 969, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python2.7/httplib.py", line 833, in _send_output
    self.send(message_body)
  File "/usr/lib/python2.7/httplib.py", line 805, in send
    self.sock.sendall(data)
  File "/opt/zato/2.0.4/eggs/requests-2.6.0-py2.7.egg/requests/packages/urllib3/contrib/pyopenssl.py", line 225, in sendall
    sent = self._send_until_done(data)
  File "/opt/zato/2.0.4/eggs/requests-2.6.0-py2.7.egg/requests/packages/urllib3/contrib/pyopenssl.py", line 215, in _send_until_done
    return self.connection.send(data)
  File "/opt/zato/2.0.4/eggs/pyOpenSSL-0.14-py2.7.egg/OpenSSL/SSL.py", line 947, in send
    raise TypeError("data must be a byte string")
~~~

**Comment by candlerb** [comment]
On a live system I reversed commit 8e9054597, but now I get a different exception:

~~~
...
  File "/opt/zato/2.0.4/eggs/requests-2.6.0-py2.7.egg/requests/packages/urllib3/connection.py", line 238, in connect
    ssl_version=resolved_ssl_version)
  File "/opt/zato/2.0.4/eggs/requests-2.6.0-py2.7.egg/requests/packages/urllib3/contrib/pyopenssl.py", line 282, in ssl_wrap_socket
    ctx.load_verify_locations(ca_certs, None)
  File "/opt/zato/2.0.4/eggs/pyOpenSSL-0.14-py2.7.egg/OpenSSL/SSL.py", line 303, in load_verify_locations
    raise TypeError("cafile must be None or a byte string")
~~~

Next I will revert to 2.0.3

**Comment by candlerb** [comment]
After rollback to 2.0.3 it is working again

**Comment by dsuch** [comment]
What OS and Python version is it on?

**Comment by candlerb** [comment]
~~~
$ cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=14.04
DISTRIB_CODENAME=trusty
DISTRIB_DESCRIPTION="Ubuntu 14.04.2 LTS"

$ python --version
Python 2.7.6
~~~

**Comment by dsuch** [comment]
The fix is in ef7771af73629c6089c0add5f037c670385ab833 and will be released in 2.0.5.

**Comment by candlerb** [comment]
What about the `cafile` parameter, and any other similar parameters?

Also is there any chance of adding a test case to catch this? It's pretty major. I can try to reproduce this in a standalone zato installation, using https to itself.

**Comment by dsuch** [comment]
Yes, it will be done. This was just to point you to a patch to cover your exact situation.

**Comment by candlerb** [comment]
For reference, I was able to replicate this on a local zato install.

## Fix the certificate

The default certificate is not usable, see #455

~~~
zato ca create server ~/env/qs-1/ca --common-name localhost qs-1 localhost
OK
~~~

## Set up SSL key+cert file

~~~
cd ~/env/qs-1/ca
cat ca/out-priv/qs-1-localhost-priv-*.pem ca/out-cert/qs-1-localhost-cert-*.pem \
    >load-balancer/config/repo/zato-lba-key-and-cert.pem
~~~

## Add to load-balancer config in GUI

~~~
frontend front_tls_optional_client_certs

  mode http
  default_backend bck_http_plain
  option forwardfor
  reqadd X-Forwarded-Proto:\ https

  acl has_x_forwarded_proto req.fhdr(X-Forwarded-Proto) -m found
  http-request deny if has_x_forwarded_proto

  bind 0.0.0.0:41223 ssl crt config/repo/zato-lba-key-and-cert.pem verify optional ca-file config/repo/zato-lba-ca-certs.pem
~~~

I also disabled SSLv3:

~~~
global
    ...
    ssl-default-bind-options no-sslv3
~~~

Validate and Save.

## Create a dummy service

I created `testing.py`

~~~
from __future__ import absolute_import, division, print_function, unicode_literals
from zato.server.service import Service

class DummyReceive(Service):
    def handle(self):
        self.logger.info("Request: %s" % self.request.raw_request)
        self.response.payload = "<ok/>"
~~~

Deploy it (for some reason it wasn't deployed when I copied to `server1/pickup-dir/`, so I deployed via web interface)

## Create incoming SOAP channel

`Connections > Channels > SOAP > new`

* Name: testing.dummy
* URL path: /testing/dummy
* Data format: XML
* Service: testing.dummy-receive
* Security definition: No security definition

## Test inbound channel

~~~
curl --cacert /opt/zato/env/qs-1/load-balancer/-ca-certs.pem \
    --data-binary '<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/"><soap:Body><foo/></soap:Body></soap:Envelope>' \
    https://localhost:41223/testing/dummy
~~~

Should respond with a soap-wrapped `<ok/>` and be recorded in the server log

## Create cert for validation

`Security > SSL/TLS > Outgoing > CA Certs > Upload a CA cert`

* Name: Sample CA
* Paste the data from `ca/ca-material/ca-cert.pem`

## Create outbound channel

`Connections > Outgoing > SOAP > new`

* Name: testing.dummy
* Host: https://localhost:41223
* URL path: /testing/dummy
* Data format: XML
* Security definition: No security definition
* TLS CA certs: Sample CA

Unfortunately, at this point clicking "ping" isn't good enough to test it, as a 500 code is returned.

## Create another service

Add to `testing.py`

~~~
class DummySend(Service):
    def handle(self):
        self.outgoing.soap.get("testing.dummy").conn.send(self.cid, "<bar/>")
        self.response.payload = "<ok/>"
~~~

Deploy it as before

## Invoke service

From the list of services, click on `testing.dummy-send`, `invoker`, `submit`.

Finally the problem is reproduced! The browser shows a long backtrace including:

~~~
...
 File "/usr/lib/python2.7/httplib.py", line 805, in send
    self.sock.sendall(data)
  File "/opt/zato/2.0.4/eggs/requests-2.6.0-py2.7.egg/requests/packages/urllib3/contrib/pyopenssl.py", line 225, in sendall
    sent = self._send_until_done(data)
  File "/opt/zato/2.0.4/eggs/requests-2.6.0-py2.7.egg/requests/packages/urllib3/contrib/pyopenssl.py", line 215, in _send_until_done
    return self.connection.send(data)
  File "/opt/zato/2.0.4/eggs/pyOpenSSL-0.14-py2.7.egg/OpenSSL/SSL.py", line 947, in send
    raise TypeError("data must be a byte string")
TypeError: data must be a byte string
~~~

## Fix

After applying the patch in ef7771a and restarting, `<ok/>` is seen as the response.

**Comment by dsuch** [comment]
The final fix is in dac375b7a4faba5373d7383f3846a2a017904d20.

Ultimately, this is all caused by SSL-related changes in 2.7.9 which meant we had to touch a few of the dependencies to support Debian 8 and this ticket is the fallout.

Turned out that it was related to all HTTP calls that were using a body payload, SOAP or not. A new test case runs a Cartesian product of all transports, data types and HTTP verbs to cover all the possible cases.

We'll need to release 2.0.5 the next week because this issue effectively blocks outgoing TLS HTTP calls.

We'll also use 2.0.5 as an opportunity to bring hotfixes back - we used to have them in 1.1 and they let us distribute patches without waiting for full releases.

@rafal-k
